Add post-initialization hooks for init command

lsviben opened the pull request:
<!--
Thank you for helping to improve Crossplane! Please read the contribution docs
(linked below) if this is your first Crossplane pull request.
-->

### Description of your changes

Adds the functionality to the`crossplane beta init` command to handle a 
NOTES.txt file (similarly to helm) and an initialization script (init.sh), which 
template repositories can use to provide more information and automate 
set-up.

The flow is following:
- clone the repo provider (no change)
- if there is a NOTES.txt file, if yes, go-template it with the provided name and print it
- if there is an init.sh script, print it and ask if to execute it. Execute it with providing it the name and directory as args

So for existing repos there wont be any change until they add a NOTES.txt, init.sh script or both.

The NOTES.txt templating for now just injects the name {{ .Name }}.

<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->

Fixes #4941 

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added or updated unit tests.
- [x] Added or updated e2e tests.
- [ ] Linked a PR or a [docs tracking issue] to [document this change].
- ~[] Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### How has this been tested

Tested locally with a repo that has [NOTES.txt](https://github.com/lsviben/function-template-go/blob/main/NOTES.txt) and [init.sh](https://github.com/lsviben/function-template-go/blob/main/init.sh) script

```
 ~/go/crossplane/_output/bin/darwin_arm64/crank beta xpkg init func https://github.com/lsviben/function-template-go
Initialized package "func" in directory "/Users/lsviben/test" from https://github.com/lsviben/function-template-go/tree/a5e76778657cd80bfe8100e77608f0577f75e291 (main)
To get started:

1. Replace `function-template-go` with func in `go.mod`,
   `package/crossplane.yaml`, and any Go imports
2. Update `input/v1beta1/` to reflect your desired input (and run `go generate`)
3. Add your logic to `RunFunction` in `fn.go`
4. Add tests for your logic in `fn_test.go`
5. Update `README.md`, to be about your function!

Found init script:
#!/bin/sh

echo hello "$1"

Do you want to run it? [y/n] y
hello func
```

lsviben commented:
Will add some unit tests for init before moving it from draft

negz commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +133,132 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+	if err != nil {
+		return errors.Wrapf(err, "failed to read notes file %s", notesFile)
+	}
+	nt, err := template.New("notes").Parse(string(ntRaw))
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse notes file %s", notesFile)
+	}
+	err = nt.Execute(k.Stdout, filler)
+	if err != nil {
+		return errors.Wrapf(err, "failed to template notes file %s", notesFile)
+	}
```
What are some reasons folks would want `NOTES.txt` to be a template?

I imagine it would explain how the extension template works, for example the `NOTES.txt` for https://github.com/crossplane/function-template-go would provide a little context on how to get started with function-template-go. I can't immediately think of a reason `NOTES.txt` would need to be templatized for that.

lsviben replied:
Its just for making the NOTES.txt a bit more personalised based on the name you chose. 
So instead of having:
```
Replace function-template-go with your function's name in go.mod, package/crossplane.yaml, and any Go imports
```
You can have:
```
Replace function-template-go with somename in go.mod, package/crossplane.yaml, and any Go imports
```

If there is a NOTES.txt with no templating in it, it will still get printed out.


TBH its just cosmetic and I dont feel to strongly about it. I can remove it for keeping the feature simpler.

negz replied:
I lean toward removing it to keep it simpler. I don't feel very strongly about it either though. 😄

lsviben replied:
Ill remove it for simplicity. It will be easier to add it later then remove.

jbw976 replied:
i'm seeing a raw `{{ .Name }}` in the output when testing now - is this because you removed the templating ability, but the NOTES.file hasn't been updated to remove the templated var?

```
❯ go run cmd/crank/main.go beta xpkg init func https://github.com/lsviben/function-template-go -d init-test
Initialized package "func" in directory "/Users/jared/dev/go/src/github.com/crossplane/crossplane/init-test" from https://github.com/lsviben/function-template-go/tree/a5e76778657cd80bfe8100e77608f0577f75e291 (main)

To get started:

1. Replace `function-template-go` with {{ .Name }} in `go.mod`,
   `package/crossplane.yaml`, and any Go imports
...
```

lsviben replied:
I just didnt update the my test function repo yet. The real NOTES.txt and init.sh that will go into the official repo will be proper.

negz commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +133,132 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+	if err != nil {
+		return errors.Wrapf(err, "failed to read init.sh file %s", scriptFile)
+	}
+	// TODO(lsviben): maybe print the script in different color to make it stand out?
+	_, err = fmt.Fprintf(k.Stdout, "\nFound init script:\n%s\n", sfRaw)
+	if err != nil {
+		return errors.Wrap(err, "failed to write to stdout")
+	}
```
What do you think about prompting the user, so that we only print out the script if they want us to?

negz commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +133,132 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+	return errors.Wrapf(cmd.Run(), "failed to run init script %s", scriptFile)
+}
+
+func promptYesNo(k *kong.Context, question string) (bool, error) {
```
What happens if someone is calling `xpkg init` from a script? It would be nice to be able to provide a flag like `--run-init-script` that would skip this prompt if true.

lsviben replied:
good idea, added

lsviben commented:
> Will add some unit tests for init before moving it from draft

Added just a test for NOTES. Dropped the test for init.sh as it has multiple prompts plus running a test script and catching its output. Maybe could add just a test without prompts if needed to make it simpler, but then its just checking if we are running a script correctly.

phisco commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +138,127 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+	}
+
+	// ask the user if they want to display the script
+	display, err := promptYesNo(k, "\nFound init.sh script. Do you want to display it?")
+	if err != nil {
+		return errors.Wrap(err, "failed to prompt user")
+	}
+	if display {
+		err = printFile(k.Stdout, scriptFile)
+		if err != nil {
+			return errors.Wrapf(err, "failed to print file %s", scriptFile)
+		}
+	}
+
+	// ask the user if they want to run the script
+	run, err := promptYesNo(k, "\nDo you want to run the script?")
+	if err != nil {
+		return errors.Wrap(err, "failed to prompt user")
+	}
+	if run {
+		return errors.Wrapf(runScript(k, scriptFile, name, dir), "failed to run init script %s", scriptFile)
+	}
+	return nil
```
Nit: what about a single prompt like `Found init.sh script. Do you want to run it? [y]es/[n]o/[v]iew`

jbw976 replied:
if they select view, what would happen after we display it? would we prompt them again?

lsviben replied:
I was thinking to do that, if they select view, just display and prompt again.

lsviben replied:
Looks like this now:
```
 ~/go/crossplane/_output/bin/darwin_arm64/crank beta xpkg init func https://github.com/lsviben/function-template-go -d func16
Initialized package "func" in directory "/tmp/t/func16" from https://github.com/lsviben/function-template-go/tree/a5e76778657cd80bfe8100e77608f0577f75e291 (main)

To get started:

1. Replace `function-template-go` with {{ .Name }} in `go.mod`,
   `package/crossplane.yaml`, and any Go imports
2. Update `input/v1beta1/` to reflect your desired input (and run `go generate`)
3. Add your logic to `RunFunction` in `fn.go`
4. Add tests for your logic in `fn_test.go`
5. Update `README.md`, to be about your function!

Found init.sh script!
Do you want to run it? [y]es/[n]o/[v]iew: v

#!/bin/sh

echo hello "$1"
Do you want to run it? [y]es/[n]o/[v]iew: y
hello func
```

The same prompt is repeated after choosing "view", in theory you could view it forever if somebody wanted, but I dont think its an issue.

jbw976 commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -60,13 +69,26 @@ following well-known template names are supported:
   # Initialize a new Provider named provider-example from a custom template.
   crossplane beta xpkg init provider-example https://github.com/crossplane/provider-template-custom
+
+  # Initialize a new Go Composition Function named function-example without
+  # prompting to display or run init.sh.
```
```suggestion
  # Initialize a new Go Composition Function named function-example and run
  # its init.sh script (if it exists) without prompting the user or displaying its contents.
```

jbw976 commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +138,127 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+
+// handleInitScript runs the init.sh script in the template repository, if it
+// exists.
+func (c *initCmd) handleInitScript(k *kong.Context, logger logging.Logger, dir, name string) error {
```
nit: looks like this function has a receiver only for access to the `RunInitScript` value. The caller is already passing in other members of the `initCmd` type, so it's mixing a bit here. Should we consider being consistent instead? (either don't pass in any members or pass them all in)

lsviben replied:
Right, missed during the cleaning. Changed it to not pass any unneeded ones plus made handleNotes the same for consistency

jbw976 commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +138,127 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+// handleNotes prints the NOTES.txt file in the template
+// repository, if it exists.
+func handleNotes(w io.Writer, logger logging.Logger, dir string) error {
+	notesFile := dir + "/" + notes
```
is https://pkg.go.dev/path/filepath#Join better to use here? same comment for the handleInitScript function too.

jbw976 reviewed (commented):
very cool @lsviben, thanks for getting this functionality added! The functionality as described in this PR mostly looks good to me! ✅ 

However, I still have a main scenario stuck in my head from our in person experience with a room of 50 contributors writing a composition function for the first time at Kubecon Chicago:

A bunch of them got confused and stuck in https://docs.crossplane.io/knowledge-base/guides/write-a-composition-function-in-go/#initialize-the-function-from-a-template, where they weren't quite sure how to go manually make all the package/import/module name changes.  They didn't do them correctly, or they missed some, etc. and their build would fail and they'd call us over to help.

I'm hoping this feature completely eliminates experiences like this. Right now, for the `function-template-go` repo, it will still print out a Notes.txt file telling them the manual steps to perform.

Can we make that better or more streamlined? Can we just do that find/replace automatically for them? 🤔 

Perhaps the right answer here is to just add a useful `init.sh` script that does this work to that repo instead of the `hello $name` script it currently has.

But, are there any lessons on the UX here that we can learn from Kubebuilder's usage of the PROJECT file? 🤔 
https://book.kubebuilder.io/cronjob-tutorial/basic-project

Possibly not because our tool (right now) is an initializer for an entire project only and does not need further functionality to continue incrementally scaffolding resources to an existing project (e.g. `kubebuilder create api ...`), but we should consider that kubebuilder UX and make sure there's nothing we should leverage. You may have already done this, but I wanted to be sure 😉

lsviben commented:
> very cool @lsviben, thanks for getting this functionality added! The functionality as described in this PR mostly looks good to me! ✅
> 
> However, I still have a main scenario stuck in my head from our in person experience with a room of 50 contributors writing a composition function for the first time at Kubecon Chicago:
> 
> A bunch of them got confused and stuck in https://docs.crossplane.io/knowledge-base/guides/write-a-composition-function-in-go/#initialize-the-function-from-a-template, where they weren't quite sure how to go manually make all the package/import/module name changes. They didn't do them correctly, or they missed some, etc. and their build would fail and they'd call us over to help.
> 
> I'm hoping this feature completely eliminates experiences like this. Right now, for the `function-template-go` repo, it will still print out a Notes.txt file telling them the manual steps to perform.
> 
> Can we make that better or more streamlined? Can we just do that find/replace automatically for them? 🤔
> 
> Perhaps the right answer here is to just add a useful `init.sh` script that does this work to that repo instead of the `hello $name` script it currently has.


The NOTES.txt and init.sh that I put in my test function-template-go repo are just to show that this feature works, not what we intend to really use in the official repos. I wanted to do it separately and after this PR is merged so not to keep changing it if we change the design. 

The goal is as you say, make it as automated as possible. So I expect the `NOTES.txt` to just say a few words about what this template is and how to use it, maybe link to some examples. And the `init.sh` to do the heavy lifting and make all the import/module/package naming changes.

Another plus is that while this change is tied to the Crossplane release cycle, adding NOTES.txt and init.sh to the template repos can be done without the need to "make it in before the release", so we can work on those as we go. Ideally after this gets merged, I would at least add them to `function-template-go` as it will be use in Contribfest. 

> 
> But, are there any lessons on the UX here that we can learn from Kubebuilder's usage of the PROJECT file? 🤔 https://book.kubebuilder.io/cronjob-tutorial/basic-project
> 
> Possibly not because our tool (right now) is an initializer for an entire project only and does not need further functionality to continue incrementally scaffolding resources to an existing project (e.g. `kubebuilder create api ...`), but we should consider that kubebuilder UX and make sure there's nothing we should leverage. You may have already done this, but I wanted to be sure 😉

I can check that, if you already have some ideas, please let me know.

jbw976 commented:
> The goal is as you say, make it as automated as possible.

Awesome @lsviben, we sound totally aligned in the goal here then! looking forward to an awesome `init.sh` script for the golang template repo after we get this functionality merged in 😉 

One thing I'm not sure of yet is how an `init.sh` script should take in arguments from the user to further refine the customization, or if they need to. Have you thought about that? 🤔 

> I can check that, if you already have some ideas, please let me know.

I don't have ideas myself yet, but I did like the pattern so it's worth checking out.  It may actually be related to the question above about the experience we should provide/recommend for `init.sh` scripts taking arguments from the user.

jbw976 reviewed (approved):
Thanks for incorporating the feedback, this is looking good to me dude! ✅ 💪

negz commented on cmd/crank/beta/xpkg/init.go:
```diff
@@ -116,14 +138,128 @@ func (c *initCmd) Run(k *kong.Context, logger logging.Logger) error {
+		return errors.Wrap(err, "failed to write to stdout")
+	}
+
+	err = c.handleNotes(k.Stdout, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to handle NOTES.txt")
+	}
```
```suggestion
	if err := c.handleNotes(k.Stdout, logger); err != nil {
		return errors.Wrap(err, "failed to handle NOTES.txt")
	}
```

Nit: In a few places you could limit the error scope.

https://github.com/crossplane/crossplane/tree/master/contributing#scope-errors

lsviben commented:
> > The goal is as you say, make it as automated as possible.
> 
> Awesome @lsviben, we sound totally aligned in the goal here then! looking forward to an awesome `init.sh` script for the golang template repo after we get this functionality merged in 😉
> 
> One thing I'm not sure of yet is how an `init.sh` script should take in arguments from the user to further refine the customization, or if they need to. Have you thought about that? 🤔

Updated the script in my function-go-template. Now it replaces all that it needs plus shows how you could take an argument if you would like to make the script do something more (ill remove that)

```
Found init.sh script!
Do you want to run it? [y]es/[n]o/[v]iew: v

#!/bin/sh

# This script helps initialize a new function project by
# replacing all instances of function-template-go with the
# name of your function. The scripts accepts two arguments:
# 1. The name of your function
# 2. The path to your function directory

set -e

cd "$2" || return

# Replace function-template-go with the name of your function
# in go.mod
perl -pi -e s,function-template-go,"$1",g go.mod
# in fn.go
perl -pi -e s,function-template-go,"$1",g fn.go
# in examples
perl -pi -e s,function-template-go,"$1",g example/*

printf 'Please provide user input: '
read -r answer

echo "You provided: $answer"
Do you want to run it? [y]es/[n]o/[v]iew: y
Please provide user input: input
You provided: input

```
